### ISSUE TITLE
Annotated types clash with class attribute names

### ISSUE THREAD

**Issue by layday** [microsoft/pyright#882]
**Describe the bug**
Types from method annotations are resolved in the context of the class and they are along with synonymous class attributes reported to be unbound.  This is not dissimilar to #863; here, however, the attributes `self.str` and `cls.str` are defined in a different scope.

**To Reproduce**
Type check the snippet below.

**Expected behavior**
That the type of `Foo.bar`, `self.str` and `cls.str` would be `str`.

**Screenshots or Code**
```py
# pyright: strict

class Foo:
    bar: str

    def __init__(self, baz: str) -> None:
        self.str = baz

    @classmethod
    def from_baz(cls, baz: str) -> None:
        cls.str = baz
```

**VS Code extension or command-line**
Extension, 1.1.55

**Comment by erictraut** [comment]
Type annotations in Python are not constrained to code-flow order. They support forward references, so their definitions can occur after their usage. In your case, you are introducing a symbol called `str` into the scope of the class, and then you're referring to a type by that same name. Pyright is attempting to resolve that type by following all available assignments, and it's encountering a circular reference, since you're defining the symbol `str` using a value that has the type of `str`.

So while this behavior is unfortunate, I don't see any way to fix it without breaking many other aspects of type checking.

It's a bad practice to redefine built-in symbols in any scope, so I recommend renaming "str" to something different. If that's not possible, you can create a local alias to the built-in str via `from builtins import str as str_builtin`.

**Comment by erictraut** [comment]
To help better explain the problem, consider the following code:

```
class Foo:
    str = int

    @classmethod
    def m1(cls, baz: str) -> None:
        pass
```

Would you agree that in this case that `str` is a class-scoped type alias that is equal to `int`, and the type of parameter `baz` is an int?

Now, let's modify that snippet as follows. Rather than introducing the symbol `str` directly within the class declaration, we introduce it within method m0. In both cases, the type checker now knows that there is a symbol `str` defined within class Foo's symbol table. It needs to consider all assignments (regardless of code flow order) to determine its value.

```
class Foo:
    @classmethod
    def m0(cls, baz: str) -> None:
        cls.str = int

    @classmethod
    def m1(cls, baz: str) -> None:
        pass
```

**Comment by layday** [comment]
Thank you for explaining.  The issue for me is that in the second example `cls.str` is defined in a different scope which I would not expect to leak into the class block and I would not typically think of this as redefining a built-in - even less so if it were an instance variable.  To understand what is happening I have to adopt a different mental model where class attributes live in the same namespace as globals.  This would be fairly obvious in a language where class variables and instance variables are bare - where they aren't prefixed by 'cls' or 'self' - but Python muddles everything up by allowing class variables _only_ to appear unprefixed in the class body.  Is `str` in the annotation of `m0` a global variable or a class variable?  If it is a class variable is it always a class variable or is it only a class variable if the class has a variable with the same name?  I think you can see how this can be confusing.  If Pyright is able to detect cyclical references in classes, could the error message be improved in cases like this one?

**Comment by layday** [comment]
I assume there was nothing more to be said then.

**Comment by erictraut** [comment]
Oops, I didn't mean to close this one. I wanted to spend a little more time investigating it.

**Comment by erictraut** [comment]
Good news. I figured out how to fix the problem. I needed to track additional information about whether class/instance variable declarations were explicit (i.e. within the class scope) or implicit (assignments that target member access expressions). The fix will be in the next version.

**Comment by layday** [comment]
Excellent, thank you!

**Comment by erictraut** [comment]
This is addressed in Pyright version 1.1.57, which I just published. It will also be included in the next version of Pylance.
